Runtime/CesiumCameraController.cs

```diff
@@ -0,0 +1,1145 @@
+    /// looks right-side up.
+    /// </summary>
+    [RequireComponent(typeof(CesiumOriginShift))]
+    public class CesiumCameraController : MonoBehaviour
+    {
+        #region User-editable properties
+
+        [SerializeField]
+        private bool _enableDynamicSpeed = true;
+
+        /// <summary>
+        /// Whether to enable dynamic speed on this controller. If enabled, 
+        /// the controller's speed will change dynamically based on elevation 
+        /// and other factors.
+        /// </summary>
+        public bool enableDynamicSpeed
+        {
+            get => this._enableDynamicSpeed;
+            set => this._enableDynamicSpeed = value;
+        }
+
+        [SerializeField]
+        [Min(0.0f)]
+        private float _dynamicSpeedMinHeight = 20.0f;
+
+        /// <summary>
+        /// The minimum height where dynamic speed starts to take effect.
+        /// Below this height, the speed will be set to the object's height 
+        /// from the Earth, which makes it move slowly when it is right above a tileset.
+        /// </summary>
+        public float dynamicSpeedMinHeight
+        {
+            get => this._dynamicSpeedMinHeight;
+            set => this._dynamicSpeedMinHeight = Mathf.Max(value, 0.0f);
+        }
+
+        [SerializeField]
+        private bool _enableDynamicClippingPlanes = true;
+
+        /// <summary>
+        /// Whether to dynamically adjust the camera's clipping planes so that
+        /// the globe will not be clipped from far away. Objects that are close 
+        /// to the camera but far above the globe in space may not appear.
+        /// </summary>
+        public bool enableDynamicClippingPlanes
+        {
+            get => this._enableDynamicClippingPlanes;
+            set => this._enableDynamicClippingPlanes = value;
+        }
+
+        [SerializeField]
+        [Min(0.0f)]
+        private float _dynamicClippingPlanesMinHeight = 10000.0f;
+
+        /// <summary>
+        /// The height to start dynamically adjusting the camera's clipping planes. 
+        /// Below this height, the clipping planes will be set to their initial values.
+        /// </summary>
+        public float dynamicClippingPlanesMinHeight
+        {
+            get => this._dynamicClippingPlanesMinHeight;
+            set => this._dynamicClippingPlanesMinHeight = Mathf.Max(value, 0.0f);
+        }
+
+        [SerializeField]
+        private AnimationCurve _flyToAltitudeProfileCurve;
+
+        /// <summary>
+        /// A curve that dictates what percentage of the max altitude the camera 
+        /// should take at a given time on the curve.
+        /// </summary>
+        /// <remarks>
+        /// This curve must be kept in the 0 to 1 range on both axes. The 
+        /// <see cref="CesiumCameraController.flyToMaximumAltitudeCurve"/>
+        /// dictates the actual max altitude at each point along the curve.
+        /// </remarks>
+        public AnimationCurve flyToAltitudeProfileCurve
```

Comment:
That makes sense. I'll separate the code into a component named `CesiumFlyToController`, but feel free to suggest other names!

Reply:
During this, I came up with some UX / API questions that I wanted to think through:
- `FlyTo` had an option to make the flight interruptible by movement. It also disabled mouse input to rotate the camera when it was in flight. By separating `CesiumFlyToController` from `CesiumCameraController`, we're separating the flight behaviour from the camera movement, so these classes will need to communicate with each other.
- A public `InterruptFlight` method in `CesiumFlyToController` seems sufficient; in their own scripts, a developer can manage if / when the flights can be interrupted. But what about the other way around in `CesiumCameraController`? Or if the flight shouldn't be interrupted at all, how do you disable the input altogether?
  - Perhaps this isn't a problem since Fly-to is manually imposing positions / rotations on the camera, so it theoretically overrides it. But I wasn't sure if it was good form to _not_ disable inputs on `CesiumCameraController`.

This made me question the best user interface for `CesiumCameraController`, once `CesiumFlyToController` becomes its own component.
- Is it helpful for `CesiumCameraController` to have an interface similar to `CharacterController` where users can move the camera by script? So for example, `MoveRight()`, `MoveUp()`, `Rotate()` etc. functions. Users can write their own scripts using those functions if they need.
- But then how would you change / use / disable key bindings? Currently they're baked into `CesiumCameraController`, and I don't think we should remove them because some people won't want to write their own controllers.

So I suppose it is a matter of potential use cases, if we ever a user will want to disable or override the camera bindings. I'm probably overthinking it, but I just wanted to throw this out there before I reorganize all of this code, just in case.

I looked around some free controllers on the Asset Store, and they all do different things. One of them was an interface that you needed to inherit from, one of them had key bindings / controls baked into the script, and one of them allowed users to modify the controls via the Inspector. So it's a matter of deciding how we want to treat our controller classes based on what our users would want.

Reply:
How's this sound?

* `CesiumFlyToController` knows about and uses `CesiumCameraController`, but not the other way around. In `Awake`, `CesiumFlyToController` looks for a `CesiumCamerController` and will use it if present, but it is not required.
* Add `EnableRotation` and `EnableMovement` properties to `CesiumCameraController`.
* At the start of a flight, `CesiumFlyToController` sets `EnableRotation` to false. If flight interruption is not allowed, it also sets `EnableMovement` to false (otherwise it leaves it true).
* If a flight may be interrupted, `CesiumFlyToController` checks at each step if the position is different from what it previously set in the last step (in ECEF probably). If so, the flight is interrupted. So the `CesiumCameraController` doesn't need to know anything about flights, it just moves as normal (if movement is enabled) and this causes the flight to be interrupted. This allows flights of other random objects to get interrupted when the objects are moved explicitly.
* At the end of the flight, `EnableRotation` and `EnableMovement` are restored to their original values.

> Is it helpful for CesiumCameraController to have an interface similar to CharacterController where users can move the camera by script? So for example, MoveRight(), MoveUp(), Rotate() etc. functions. Users can write their own scripts using those functions if they need.

This sounds useful. Maybe even `MoveNorth` etc. But I think it can be a future feature because introducing it later doesn't have any backward compatibility implications.

> But then how would you change / use / disable key bindings? 

I believe in Unreal there's an abstraction here, where you can define conceptual inputs (like "move left"), and then separately and via runtime configuration bind those inputs to specific keys / controls / actions. I guess there's nothing like that in Unity? If not, I think we ultimately should allow the user to configure the bindings using the Inspector and by setting properties at runtime, but that can be a future feature (again, no backward compatibility concerns really).

Reply:
@kring This all sounds reasonable to me. Thanks for thinking it through with so much detail :pray: